### DISCUSSION TITLE
Report response time breakdown on http response using Server Timing API

### DISCUSSION THREAD

**Discussion by sarod** [glowroot/glowroot#915] [Ideas]
First of all: Glowroot is great!! So thank you for this great tool!🥇 

I was wondering how easy it would be support a mode in which glow root would inject the response time breakdown information (time spent on jdbc and such...) on http responses using the [Server Timing](https://w3c.github.io/server-timing/) format?

Additional info on how [Server Timing API is surfaced in chrome browser](https://web.dev/custom-metrics/?utm_source=devtools#server-timing-api)

**Comment by Never-lab** [comment]
Thanks for the idea — this would be a nice opt-in bridge between Glowroot’s timer tree and browser DevTools.

### Feasibility (rough take)
Glowroot already builds the breakdown you care about (e.g. JDBC timers under the servlet transaction). The natural hook is late in the servlet lifecycle (around when the servlet plugin ends the root trace entry), where those child timers usually already exist. Glowroot can *capture* response headers today, but it does not *write* its own.

A v1 could look like:
- **Opt-in** servlet plugin property (default **off**)
- Emit `Server-Timing` with a small set of metrics (e.g. top-level / whitelisted timers: `jdbc`, total request, …)
- Sanitize timer names to valid Server-Timing metric tokens
- **Best-effort only**: set the header only if the response is still mutable (not committed)

### Caveats worth deciding up front
1. **Committed responses** — streaming / early flush means headers cannot be added anymore; buffering the whole body just for this would be too heavy for v1.
2. **Async servlets** — completion paths need care so we don’t write incomplete or duplicate timings.
3. **Security / information leak** — Server-Timing is visible to clients (and often intermediaries). Must stay opt-in; maybe document “dev / trusted networks only”.
4. **javax + jakarta** servlet plugins would both need the same behavior.
5. This complements Glowroot UI / storage — it does not replace them.

I’m interested in this as a small upstream enhancement if maintainers think it fits the project scope.

cc @trask @nowheresly — would you be open to an opt-in Server-Timing export from the servlet plugin (best-effort, default off), or would you rather keep response mutation out of core plugins?
